man/dmmbctl.1
.TH "DMMBCTL" "1" "" "Version 2.0.0" "User Commands"
.SH NAME
dmmbctl \(en read or write value from or to Modbus RTU/TCP register
.SH SYNOPSIS
\f[B]dmmbctl\f[R] \-\-help
.PP
\f[B]dmmbctl\f[R] \-\-version
.PP
\f[B]dmmbctl\f[R] \-\-\f[B]path\f[R] \f[I]path\f[R]
\-\-\f[B]baudrate\f[R] \f[I]n\f[R] \-\-\f[B]bytesize\f[R] \f[I]n\f[R]
\-\-\f[B]parity\f[R] \f[I]name\f[R] \-\-\f[B]stopbits\f[R] \f[I]n\f[R]
\-\-\f[B]slave\f[R] \f[I]n\f[R] \-\-\f[B]type\f[R] \f[I]name\f[R]
[\-\-\f[B]order\f[R] \f[I]name\f[R]] [\-\-\f[B]code\f[R] \f[I]n\f[R]]
[\-\-\f[B]read\f[R] \f[I]address\f[R]] [\-\-\f[B]write\f[R]
\f[I]address\f[R]] [\-\-\f[B]value\f[R] \f[I]n\f[R]]
.PP
\f[B]dmmbctl\f[R] \-\-\f[B]address\f[R] \f[I]ip\f[R] \-\-\f[B]port\f[R]
\f[I]port\f[R] \-\-\f[B]slave\f[R] \f[I]n\f[R] \-\-\f[B]type\f[R]
\f[I]name\f[R] [\-\-\f[B]order\f[R] \f[I]name\f[R]] [\-\-\f[B]code\f[R]
\f[I]n\f[R]] [\-\-\f[B]read\f[R] \f[I]address\f[R]] [\-\-\f[B]write\f[R]
\f[I]address\f[R]] [\-\-\f[B]value\f[R] \f[I]n\f[R]]
[\-\-\f[B]debug\f[R]]
.SH DESCRIPTION
The \f[B]dmmbctl\f[R] command\-line program reads a value from or writes
a value to a register of a connected Modbus RTU/TCP device.
Modbus RTU requires the command\-line arguments \f[CR]\-\-path\f[R],
\f[CR]\-\-baudrate\f[R], \f[CR]\-\-bytesize\f[R], \f[CR]\-\-parity\f[R],
and \f[CR]\-\-stopbits\f[R].
For Modbus TCP, only \f[CR]\-\-address\f[R] and \f[CR]\-\-port\f[R] must
be passed.
.PP
The following data types are supported:
.TP
\f[CR]int16\f[R]
2\-byte signed integer.
.TP
\f[CR]int32\f[R]
4\-byte signed integer.
.TP
\f[CR]uint16\f[R]
2\-byte unsigned integer.
.TP
\f[CR]uint32\f[R]
4\-byte unsigned integer.
.TP
\f[CR]float\f[R]
4\-byte float.
.PP
In order to read floating\-point values, set \f[CR]\-\-type\f[R] to
\f[CR]float\f[R] and \f[CR]\-\-order\f[R] to the byte order used by the
Modbus device, either \f[CR]abcd\f[R], \f[CR]badc\f[R], \f[CR]cdab\f[R],
or \f[CR]dcba\f[R].
Only integer values may be written to a register.
.SH OPTIONS
.TP
\f[B]\-\-address\f[R], \f[B]\-a\f[R] \f[I]ip\f[R]
Modbus TCP address (IPv4).
.TP
\f[B]\-\-baudrate\f[R], \f[B]\-B\f[R] \f[I]n\f[R]
Modbus RTU baud rate (9600, 19200, \&...).
.TP
\f[B]\-\-bytesize\f[R], \f[B]\-Z\f[R] \f[I]n\f[R]
Modbus RTU byte size (5, 6, 7, 8).
.TP
\f[B]\-\-code\f[R], \f[B]\-C\f[R] \f[I]n\f[R]
Optional Modbus function code.
Set to 1 to read coil status (function 0x01) and to 5 to write coil
status (function 0x05).
.TP
\f[B]\-\-debug\f[R], \f[B]\-V\f[R]
Print debug messages from \f[I]libmodbus\f[R].
.TP
\f[B]\-\-help\f[R], \f[B]\-h\f[R]
Print available command\-line arguments and quit.
.TP
\f[B]\-\-order\f[R], \f[B]\-b\f[R] [abcd|badc|cdab|dcba]
Byte order of float (\f[CR]abcd\f[R], \f[CR]badc\f[R], \f[CR]cdab\f[R],
\f[CR]dcba\f[R]).
.TP
\f[B]\-\-parity\f[R], \f[B]\-P\f[R] [none|even|odd]
Modbus RTU parity bits (\f[CR]none\f[R], \f[CR]even\f[R],
\f[CR]odd\f[R]).
.TP
\f[B]\-\-path\f[R], \f[B]\-p\f[R] \f[I]path\f[R]
Modbus RTU device path.
.TP
\f[B]\-\-port\f[R], \f[B]\-q\f[R] \f[I]port\f[R]
Modbus TCP port.
.TP
\f[B]\-\-read\f[R], \f[B]\-r\f[R] \f[I]register\f[R]
Read value from given Modbus register address.
.TP
\f[B]\-\-slave\f[R], \f[B]\-s\f[R] \f[I]n\f[R]
Slave id of Modbus device.
.TP
\f[B]\-\-stopbits\f[R], \f[B]\-O\f[R] \f[I]n\f[R]
Modbus RTU stop bits (1, 2).
.TP
\f[B]\-\-type\f[R], \f[B]\-t\f[R] [int16|int32|uint16|uint32|float]
Number type (\f[CR]int16\f[R], \f[CR]int32\f[R], \f[CR]uint16\f[R],
\f[CR]uint32\f[R], \f[CR]float\f[R]).
.TP
\f[B]\-\-value\f[R], \f[B]\-i\f[R] \f[I]n\f[R]
Integer value to write.
.TP
\f[B]\-\-version\f[R], \f[B]\-v\f[R]
Print version information and quit.
.TP
\f[B]\-\-write\f[R] , \f[B]\-w\f[R] \f[I]register\f[R]
Write value to given Modbus register address.
.SH EXIT STATUS
.TP
\f[B]0\f[R]
Success.
Modbus access was successful.
.TP
\f[B]1\f[R]
Failure.
Program execution failed.
.SH EXAMPLE
Read the current temperature in degrees Celsius measured by a Pt100 RTD
that is connected to an I/O module with Modbus RTU interface:
.IP
.EX
$ dmmbctl \-p /dev/ttyUSB0 \-B 19200 \-Z 8 \-P even \-O 1 \(rs
  \-s 1 \-r 50 \-t float \-b abcd
21.217552185059
.EE
.PP
The I/O module is attached through an RS\-485 adapter on
\f[CR]/dev/ttyUSB\f[R] (19200 baud, 8E1) and configured to use slave id
1.
The value is read from register 50 and converted to float in
\f[CR]abcd\f[R] byte order.
.SH SEE ALSO
\f[I]dmmb(1)\f[R]


CHANGELOG.md
# Change Log

All notable changes to the project will be documented in this file.

## [2.0.0] – Unreleased

## Library

* `Added` error codes
* `Added` _fortran-nng_ interface bindings to NNG to `vendor/`
* `Added` module `dm_arg_parser` (outsourced from `dm_arg`)
* `Added` module `dm_group` for observation groups
* `Added` module `dm_ipc` for socket access
* `Added` module `dm_ipc_async` for asynchronous I/O
* `Added` module `dm_ipc_disco` for service discovery
* `Added` module `dm_ipc_message` for message handling
* `Added` module `dm_ipc_mutex` for mutex access
* `Added` module `dm_ipc_thread` for thread access
* `Added` module `dm_ipc_type` of IPC enumerators
* `Added` module `dm_job_list` (outsourced from `dm_job`)
* `Changed` names of modules `dm_fifo`, `dm_mqueue`, `dm_mqueue_util`, `dm_pipe`,
  `dm_sem`, `dm_signal`, `dm_thread`, `dm_tty` to include `posix` prefix
* `Changed` name of module `dm_system` to `dm_posix`
* `Changed` database schema due to new data models
* `Changed` default database suffix from `.sqlite` to `.db`
* `Changed` GeoCOM API for Lua to accept prototype observation as first argument
* `Changed` model of observation data structure `dm_observ`
* `Deleted` model of request data strucutre `dm_request`

## Programs

* `Changed` response value units in `dmved`

## Documentation

* Converted user guide from AsciiDoc to Markdown
* Converted man pages from AsciiDoc to Markdown

## [1.0.0] – Unreleased

## Library

* `Added` Bessel, Butterworth, Chebyshev filters to `dm_filter`

## [0.9.9] – 2026-01-11

### Library

* `Added` module `dm_coord` for coordinate transformation
* `Added` module `dm_filter` for filtering
* `Added` module `dm_ghostscript` for Ghostscript access
* `Added` module `dm_netstring` for netstring parsing
* `Added` module `dm_statistics` of statistics functions
* `Added` file tree size function to module `dm_file`
* `Added` file modification date/time to function `dm_file_touch()` in `dm_file`
* `Added` format `FORMAT_TSV` for tab-separated values to `dm_format`
* `Added` swap routines to `dm_util`
* `Changed` error code order for future additions
* `Changed` default database cache size of `dm_db_open()` to improve performance

### Programs

* `Added` author option to `dmreport`
* `Added` meta data to PDF output of `dmreport`
* `Fixed` reading of options in `dmdbctl`
* `Changed` last modified date/time to last log date/time in output of `dmfeed`

## [0.9.8] – 2025-08-30

### Library

* `Added` module `dm_image` for image handling
* `Added` module `dm_js` for JavaScript generation
* `Added` module `dm_roff` for GNU roff abstraction
* `Added` module `dm_serial` for derived type serialisation
* `Added` module `dm_transfer` for HTTP-RPC API file transfers
* `Added` GeoJSON Feature Collection procedures to `dm_geojson`
* `Added` image upload to `dm_rpc`
* `Added` HTTP request and response headers to `dm_rpc`
* `Added` HTTP response headers to `dm_fcgi`
* `Added` terminals `gpic`, `postscript`, and `sixeltek` to `dm_plot`
* `Added` image and transfer functions to `dm_db_api`
* `Changed` structure of database abstraction layer, added modules `dm_db_api`, `dm_db_count`, `dm_db_json`, `dm_db_pragma`, and `dm_db_row`
* `Changed` API of modules `dm_arg` and `dm_config` to object-oriented
* `Changed` query parameter API in `dm_cgi`
* `Changed` Git submodules to copies of third-party dependencies
* `Fixed` reading from pipe in `dm_pipe`
* `Fixed` unsigned type conversion in `dm_c`
* `Fixed` return code handling of database select functions

### Programs

* `Added` program `dmcamera` for capturing of camera images
* `Added` program `dmupload` for image upload to HTTP-RPC API
* `Added` scale factor for response values to `dmreport`
* `Added` PDF and PostScript output to `dmreport`
* `Added` message queue reading to `dmmb`
* `Added` image upload to `dmapi`
* `Added` image viewer to `dmweb`
* `Added` environment variables `DM_IMAGE_DB`, `DM_IMAGE_DIR` to `dmapi` and `dmweb`
* `Added` log file output to `dmlogger`
* `Changed` environment variables in `dmapi` and `dmweb` to `DM_BEAT_DB`, `DM_LOG_DB`, `DM_OBSERV_DB`
* `Changed` derived type serialisation in `dmapi` and `dmexport`
* `Changed` database access in `dmexport`
* `Changed` configuration of `dmgrc`

## [0.9.7] – 2025-04-05

### Library

* `Added` module `dm_c` for C interoperability
* `Added` module `dm_camera` for webcam and RTSP access (FFmpeg)
* `Added` module `dm_crypto` of cryptographic hash functions
* `Added` module `dm_db_query` for SQL query building
* `Added` module `dm_dwd` for DWD API format parsing
* `Added` module `dm_dwd_api` for DWD API access (libcurl)
* `Added` module `dm_ftp` for FTP transfer (libcurl)
* `Added` module `dm_freebsd` for system status access on FreeBSD
* `Added` module `dm_geojson` for GeoJSON serialisation
* `Added` module `dm_gm` for GraphicsMagick access
* `Added` module `dm_im` for XMPP connectivity (libstrophe)
* `Added` module `dm_linux` for system status access on Linux
* `Added` module `dm_modbus` for Modbus RTU/TCP connectivity (libmodbus)
* `Added` module `dm_ve` for Victron Energy VE.Direct protocol support
* `Added` library build date and system type parameters (CPP, FPP)
* `Added` Linux and FreeBSD abstractions in module `dm_system`

app/dmmbctl.f90
! dmmbctl.f90
!
! Author:  Philipp Engel
! Licence: ISC
program dmmbctl
    !! Basic command-line tool for Modbus RTU/TCP communication.
    use :: dmpack
    implicit none (type, external)

    character(*), parameter :: APP_NAME  = 'dmmbctl'
    integer,      parameter :: APP_MAJOR = 2
    integer,      parameter :: APP_MINOR = 0
    integer,      parameter :: APP_PATCH = 0

    type :: app_rtu_type
        !! Modbus RTU settings.
        character(FILE_PATH_LEN) :: path      = ' '                   !! Path.
        integer                  :: baud_rate = POSIX_TTY_B19200      !! Baud rate.
        integer                  :: byte_size = POSIX_TTY_BYTE_SIZE8  !! Byte size.
        integer                  :: parity    = POSIX_TTY_PARITY_EVEN !! Parity name.
        integer                  :: stop_bits = POSIX_TTY_STOP_BITS1  !! Stop bits.
    end type app_rtu_type

    type :: app_tcp_type
        !! Modbus TCP settings.
        character(NET_IPV4_LEN) :: address = ' ' !! IPv4 address.
        integer                 :: port    = 0   !! Port.
    end type app_tcp_type

    type :: app_type
        !! Application settings.
        integer            :: mode     = MODBUS_MODE_NONE    !! Modbus mode (RTU, TCP).
        integer            :: slave    = 1                   !! Modbus slave id.
        integer            :: register = 0                   !! Modbus register address.
        integer            :: access   = MODBUS_ACCESS_NONE  !! Read or write operation.
        integer            :: code     = int(z'00')          !! Modbus function code.
        integer            :: type     = MODBUS_TYPE_DEFAULT !! Number type.
        integer            :: order    = MODBUS_ORDER_NONE   !! Byte order of type float.
        integer            :: value    = 0                   !! Value to write.
        logical            :: debug    = .false.             !! Enable debug mode of libmodbus.
        type(app_rtu_type) :: rtu                            !! Modbus RTU settings.
        type(app_tcp_type) :: tcp                            !! Modbus TCP settings.
    end type app_type

    integer        :: rc  ! Return code.
    type(app_type) :: app ! App settings.

    ! Initialise DMPACK.
    call dm_init()

    ! Get command-line arguments and read options from configuration file.
    rc = read_args(app)
    if (dm_is_error(rc)) call dm_stop(STOP_FAILURE)

    ! Read from or write to Modbus address.
    rc = run(app)
    if (dm_is_error(rc)) call dm_stop(STOP_FAILURE)
contains
    integer function run(app) result(rc)
        type(app_type), intent(inout) :: app !! App type.

        type(modbus_rtu_type), target  :: modbus_rtu ! Modbus RTU type.
        type(modbus_tcp_type), target  :: modbus_tcp ! Modbus TCP type.
        class(modbus_type),    pointer :: modbus     ! Modbus pointer.

        modbus_block: block
            rc = E_INVALID

            ! Create Modbus RTU/TCP context.
            select case (app%mode)
                case (MODBUS_MODE_RTU)
                    rc = dm_modbus_create(modbus    = modbus_rtu,        &
                                          path      = app%rtu%path,      &
                                          baud_rate = app%rtu%baud_rate, &
                                          byte_size = app%rtu%byte_size, &
                                          parity    = app%rtu%parity,    &
                                          stop_bits = app%rtu%stop_bits)
                    modbus => modbus_rtu
                case (MODBUS_MODE_TCP)
                    rc = dm_modbus_create_tcp(modbus_tcp, app%tcp%address, app%tcp%port)
                    modbus => modbus_tcp
            end select

            if (dm_is_error(rc)) then
                call dm_error_out(rc, 'failed to create Modbus context: ' // dm_modbus_error_message())
                exit modbus_block
            end if

            ! Debug mode.
            if (app%debug) then
                rc = dm_modbus_set_debug(modbus, .true.)

                if (dm_is_error(rc)) then
                    call dm_error_out(rc, 'failed to enable debug mode: ' // dm_modbus_error_message())
                    exit modbus_block
                end if
            end if

            ! Create Modbus connection.
            rc = dm_modbus_connect(modbus)

            if (dm_is_error(rc)) then
                call dm_error_out(rc, 'failed to create Modbus connection: ' // dm_modbus_error_message())
                exit modbus_block
            end if

            ! Set slave device.
            rc = dm_modbus_set_slave(modbus, slave=app%slave)

            if (dm_is_error(rc)) then
                call dm_error_out(rc, 'failed to set slave id to ' // dm_itoa(app%slave))
                exit modbus_block
            end if

            ! Read or write value.
            select case (app%access)
                case (MODBUS_ACCESS_READ)
                    rc = read_value(modbus, app%register, app%code, app%type, app%order)
                    if (dm_is_error(rc)) call dm_error_out(rc, 'failed to read ' // MODBUS_TYPE_NAMES(app%type))
                case (MODBUS_ACCESS_WRITE)
                    rc = write_value(modbus, app%register, app%code, app%type, app%value)
                    if (dm_is_error(rc)) call dm_error_out(rc, 'failed to write ' // MODBUS_TYPE_NAMES(app%type))
            end select
        end block modbus_block

        ! Disconnect and clean-up.
        call dm_modbus_close(modbus)
        call dm_modbus_destroy(modbus)
    end function run

    integer function read_value(modbus, address, code, type, order) result(rc)
        !! Reads and prints value from register.
        character(*), parameter :: FMT_INT  = '(i0)'   !! Integer output format.
        character(*), parameter :: FMT_REAL = '(f0.8)' !! Real output format.

        class(modbus_type), intent(inout) :: modbus  !! Modbus RTU/TCP type.
        integer,            intent(in)    :: address !! Register address.
        integer,            intent(in)    :: code    !! Modbus function code or 0.
        integer,            intent(in)    :: type    !! Type of value.
        integer,            intent(in)    :: order   !! Byte order.

        integer(i2) :: i16
        integer(i4) :: i32
        integer(i8) :: i64
        real(r4)    :: r32

        ! Read and output coil status (0x01).
        if (code == int(z'01')) then
            rc = dm_modbus_read_bit(modbus, address, i32)
            print FMT_INT, i32
            return
        end if

        ! Read from input (0x04) or holding (0x03) register.
        select case (type)
            case (MODBUS_TYPE_INT16);  rc = dm_modbus_read_int16 (modbus, address, i16)
            case (MODBUS_TYPE_INT32);  rc = dm_modbus_read_int32 (modbus, address, i32)
            case (MODBUS_TYPE_UINT16); rc = dm_modbus_read_uint16(modbus, address, i32)
            case (MODBUS_TYPE_UINT32); rc = dm_modbus_read_uint32(modbus, address, i64)
            case (MODBUS_TYPE_FLOAT);  rc = dm_modbus_read_float (modbus, address, r32, order)
            case default;              rc = E_INVALID
        end select

        if (dm_is_error(rc)) return

        ! Output value.
        select case (type)
            case (MODBUS_TYPE_INT16);  print FMT_INT,  i16
            case (MODBUS_TYPE_INT32);  print FMT_INT,  i32
            case (MODBUS_TYPE_UINT16); print FMT_INT,  i32
            case (MODBUS_TYPE_UINT32); print FMT_INT,  i64
            case (MODBUS_TYPE_FLOAT);  print FMT_REAL, r32
        end select
    end function read_value

    integer function write_value(modbus, address, code, type, value) result(rc)
        !! Writes value to register.
        class(modbus_type), intent(inout) :: modbus  !! Modbus RTU/TCP type.
        integer,            intent(in)    :: address !! Register address.
        integer,            intent(in)    :: code    !! Modbus function code or 0.
        integer,            intent(in)    :: type    !! Type of value.
        integer,            intent(in)    :: value   !! Value to write.

        ! Write to coil (0x05).
        if (code == int(z'05')) then
            rc = dm_modbus_write_bit(modbus, address, value)
            return
        end if

        ! Write to holding register (0x06).
        select case (type)
            case (MODBUS_TYPE_INT16);  rc = dm_modbus_write_int16 (modbus, address, int(value, kind=i2))
            case (MODBUS_TYPE_INT32);  rc = dm_modbus_write_int32 (modbus, address, value)
            case (MODBUS_TYPE_UINT16); rc = dm_modbus_write_uint16(modbus, address, value)
            case (MODBUS_TYPE_UINT32); rc = dm_modbus_write_uint32(modbus, address, int(value, kind=i8))
            case default;              rc = E_INVALID
        end select
    end function write_value

    ! **************************************************************************
    ! COMMAND-LINE ARGUMENTS.
    ! **************************************************************************
    integer function read_args(app) result(rc)
        !! Reads command-line arguments and settings from configuration file.
        type(app_type), intent(out) :: app !! App type.

        logical :: has_read, has_write
        logical :: has_order, has_type, has_value
        logical :: has_baud_rate, has_byte_size, has_parity, has_path, has_stop_bits
        logical :: has_address, has_port

        type(arg_parser_class) :: parser

        call parser%add('read',      short='r', type=ARG_TYPE_INTEGER)                                  ! -r, --read <register>
        call parser%add('write',     short='w', type=ARG_TYPE_INTEGER)                                  ! -w, --write <register>
        call parser%add('path',      short='p', type=ARG_TYPE_FILE)                                     ! -p, --path <path>
        call parser%add('baudrate',  short='B', type=ARG_TYPE_INTEGER)                                  ! -B, --baudrate <n>
        call parser%add('bytesize',  short='Z', type=ARG_TYPE_INTEGER)                                  ! -Z, --bytesize <n>
        call parser%add('parity',    short='P', type=ARG_TYPE_STRING)                                   ! -P, --parity <string>
        call parser%add('stopbits',  short='O', type=ARG_TYPE_INTEGER)                                  ! -O, --stopbits <n>
        call parser%add('address',   short='a', type=ARG_TYPE_STRING,  min_len=7, max_len=NET_IPV4_LEN) ! -a, --address <string>
        call parser%add('port',      short='q', type=ARG_TYPE_INTEGER)                                  ! -q, --port <n>
        call parser%add('slave',     short='s', type=ARG_TYPE_INTEGER, required=.true.)                 ! -s, --slave <n>
        call parser%add('code',      short='C', type=ARG_TYPE_INTEGER)                                  ! -C, --code <n>
        call parser%add('type',      short='t', type=ARG_TYPE_STRING,  min_len=5, max_len=6)            ! -t, --type <string>
        call parser%add('order',     short='b', type=ARG_TYPE_STRING,  min_len=4, max_len=4)            ! -b, --order <string>
        call parser%add('value',     short='i', type=ARG_TYPE_INTEGER)                                  ! -i, --value <n>
        call parser%add('debug',     short='D', type=ARG_TYPE_LOGICAL)                                  ! -D, --debug

        ! Read all command-line arguments.
        rc = parser%read(version_callback)
        if (dm_is_error(rc)) return

        block
            character(4) :: order, parity
            character(6) :: type
            integer      :: baud_rate, byte_size, stop_bits
            integer      :: read_register, write_register

            call parser%get('read',     read_register,   passed=has_read)
            call parser%get('write',    write_register,  passed=has_write)
            call parser%get('path',     app%rtu%path,    passed=has_path)
            call parser%get('baudrate', baud_rate,       passed=has_baud_rate)
            call parser%get('bytesize', byte_size,       passed=has_byte_size)
            call parser%get('parity',   parity,          passed=has_parity)
            call parser%get('stopbits', stop_bits,       passed=has_stop_bits)
            call parser%get('address',  app%tcp%address, passed=has_address)
            call parser%get('port',     app%tcp%port,    passed=has_port)
            call parser%get('slave',    app%slave)
            call parser%get('code',     app%code)
            call parser%get('type',     type,            passed=has_type)
            call parser%get('order',    order,           passed=has_order)
            call parser%get('value',    app%value,       passed=has_value)
            call parser%get('debug',    app%debug)

            if (has_baud_rate) app%rtu%baud_rate = dm_posix_tty_baud_rate_from_value(baud_rate)
            if (has_byte_size) app%rtu%byte_size = dm_posix_tty_byte_size_from_value(byte_size)
            if (has_parity)    app%rtu%parity    = dm_posix_tty_parity_from_name(parity)
            if (has_stop_bits) app%rtu%stop_bits = dm_posix_tty_stop_bits_from_value(stop_bits)
            if (has_type)      app%type          = dm_modbus_type_from_name(type)
            if (has_order)     app%order         = dm_modbus_order_from_name(order)

            if (has_path) then
                app%mode = MODBUS_MODE_RTU
            else
                app%mode = MODBUS_MODE_TCP
            end if

            if (has_read) then
                ! Modbus register to read from.
                app%access   = MODBUS_ACCESS_READ
                app%register = read_register
            else
                ! Modbus register to write to.
                app%access   = MODBUS_ACCESS_WRITE
                app%register = write_register
            end if
        end block

        ! Validate settings.
        rc = E_INVALID

        ! Path and address.
        if (.not. has_path .and. .not. has_address) then
            call dm_error_out(rc, 'argument --path or --address required')
            return
        end if

        if (has_path .and. has_address) then
            call dm_error_out(rc, 'argument --path conflicts with --address')
            return
        end if

        ! Modbus RTU.
        if (app%mode == MODBUS_MODE_RTU) then
            if (.not. has_baud_rate) then
                call dm_error_out(rc, 'argument --baudrate is required')
                return
            end if

            if (.not. has_byte_size) then
                call dm_error_out(rc, 'argument --bytesize is required')
                return
            end if

            if (.not. has_parity) then
                call dm_error_out(rc, 'argument --parity is required')
                return
            end if

            if (.not. has_stop_bits) then
                call dm_error_out(rc, 'argument --stopbits is required')
                return
            end if

            ! Invalid arguments.
            if (has_port) then
                call dm_error_out(rc, 'argument --port is not permitted')
                return
            end if

            ! TTY path.
            if (.not. dm_file_exists(app%rtu%path)) then
                rc = E_NOT_FOUND
                call dm_error_out(rc, 'device ' // trim(app%rtu%path) // ' not found')
                return
            end if

            ! TTY baud rate.
            if (.not. dm_posix_tty_baud_rate_is_valid(app%rtu%baud_rate)) then
                call dm_error_out(rc, 'argument --baudrate is invalid')
                return
            end if

            ! TTY byte size.
            if (.not. dm_posix_tty_byte_size_is_valid(app%rtu%byte_size)) then
                call dm_error_out(rc, 'argument --bytesize is invalid')
                return
            end if

            ! TTY parity.
            if (.not. dm_posix_tty_parity_is_valid(app%rtu%parity)) then
                call dm_error_out(rc, 'argument --parity is invalid')
                return
            end if

            ! TTY stop bits.
            if (.not. dm_posix_tty_stop_bits_is_valid(app%rtu%stop_bits)) then
                call dm_error_out(rc, 'argument --stopbits is invalid')
                return
            end if
        end if

        ! Modbus TCP.
        if (app%mode == MODBUS_MODE_TCP) then
            if (.not. has_port) then
                call dm_error_out(rc, 'argument --port is required')
                return
            end if

            ! Invalid arguments.
            if (has_baud_rate) then
                call dm_error_out(rc, 'argument --baudrate is not permitted')
                return
            end if

            if (has_byte_size) then
                call dm_error_out(rc, 'argument --bytesize is not permitted')
                return
            end if

            if (has_parity) then
                call dm_error_out(rc, 'argument --parity is not permitted')
                return
            end if

            if (has_stop_bits) then
                call dm_error_out(rc, 'argument --stopbits is not permitted')
                return
            end if

            ! Modbus TCP options.
            if (.not. dm_net_ipv4_is_valid(app%tcp%address)) then
                call dm_error_out(rc, 'argument --address is not a valid IPv4 address')
                return
            end if

            if (app%tcp%port < 1) then
                call dm_error_out(rc, 'argument --port must be > 0')
                return
            end if
        end if

        ! Slave id.
        if (app%slave < 1) then
            call dm_error_out(rc, 'argument --slave must be > 0')
            return
        end if

        ! Read xor write access.
        if (has_read .and. has_write) then
            call dm_error_out(rc, 'argument --read conflicts with --write')
            return
        else if (.not. has_read .and. .not. has_write) then
            call dm_error_out(rc, 'argument --read or --write is missing')
            return
        end if

        if (has_read) then
            ! Modbus register to read from.
            if (app%register < 0) then
                call dm_error_out(rc, 'argument --read is not a valid Modbus register')
                return
            end if
        else
            ! Modbus register to write to.
            if (app%register < 0) then
                call dm_error_out(rc, 'argument --write is not a valid Modbus register')
                return
            end if

            if (.not. has_value) then
                call dm_error_out(rc, 'argument --write requires --value')
                return
            end if
        end if

        ! Modbus function code.
        if (app%code < 0) then
            call dm_error_out(rc, 'argument --code is invalid')
            return
        end if

        ! Number type (int16, int32, uint16, uint32, float).
        if (has_type) then
            if (.not. dm_modbus_type_is_valid(app%type)) then
                call dm_error_out(rc, 'argument --type is not a valid number type')
                return
            end if

            if (app%type == MODBUS_TYPE_FLOAT .and. app%access == MODBUS_ACCESS_WRITE) then
                call dm_error_out(rc, 'argument --write is not allowed for type float')
                return
            end if
        end if

        ! Byte order.
        if (has_order) then
            if (app%type /= MODBUS_TYPE_FLOAT .and. app%order /= MODBUS_ORDER_NONE) then
                call dm_error_out(rc, 'argument --order is not allowed for type ' // MODBUS_TYPE_NAMES(app%type))
                return
            end if

            if (.not. dm_modbus_order_is_valid(app%order)) then
                call dm_error_out(rc, 'argument --order is not a valid byte order')
                return
            end if
        end if

        rc = E_NONE
    end function read_args

    ! **************************************************************************
    ! CALLBACKS.
    ! **************************************************************************
    subroutine version_callback()
        call dm_version_out(APP_NAME, APP_MAJOR, APP_MINOR, APP_PATCH)
        print '(a)', dm_modbus_version(.true.)
    end subroutine version_callback
end program dmmbctl
